tests/std/tests/P0040R3_parallel_memory_algorithms/test.cpp

```diff
@@ -0,0 +1,418 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <atomic>
+#include <cassert>
+#include <execution>
+#include <iterator>
+#include <memory>
+#include <numeric>
+
+#include <parallel_algorithms_utilities.hpp>
+
+using namespace std;
+using namespace std::execution;
+
+class wrap_uchar {
+private:
+    unsigned char ch_ = 42;
+
+public:
+    bool is_expected() const noexcept {
+        return ch_ == 42;
+    }
+};
+
+static_assert(sizeof(wrap_uchar) == 1);
+
+const auto expectation = [](const wrap_uchar& x) { return x.is_expected(); };
+
+const auto expectation_zero = [](int n) { return n == 0; };
+
+constexpr auto bad_uchar = static_cast<unsigned char>(0xcd);
+constexpr auto bad_int   = static_cast<int>(0xdeadbeaf);
+
+struct resetting_guard {
+    int* ptr_ = nullptr;
+
+    ~resetting_guard() {
+        if (ptr_) {
+            *ptr_ = 0;
+        }
+    }
+
+    resetting_guard& operator=(const resetting_guard&) = delete;
+    resetting_guard& operator=(resetting_guard&&)      = delete;
+};
+
+template <class T>
+struct deallocating_only_deleter {
+    size_t count_;
+
+    void operator()(T* ptr) const noexcept {
+        allocator<T>{}.deallocate(ptr, count_);
+    }
+};
+
+template <class T>
+unique_ptr<T, deallocating_only_deleter<T>> make_constructed_nondestroying_buffer(size_t n) {
+    if (n == 0) {
+        return unique_ptr<T, deallocating_only_deleter<T>>{};
+    }
+
+    allocator<T> al;
+    auto up = unique_ptr<T, deallocating_only_deleter<T>>{al.allocate(n), deallocating_only_deleter<T>{n}};

```

Comment:
Style: This is using the `auto var = Type{init};` pattern that we conventionally avoid. It would be simpler to directly specify `up`'s type.